File: Programmers/Lv.2/프린터.py
def solution(priorities, location):
    answer = 0
    
    while len(priorities) != 0:
        if location == 0: # 요청 문서가 맨 앞에 위치
            if priorities[0] < max(priorities): # 더 중요도가 큰 문서 존재 시
                priorities.append(priorities.pop(0)) # 맨 뒤 이동 후
                location = len(priorities) - 1 # 요청 문서 위치 맨 뒤로 조정
            else:
                return answer + 1
        else:
            if priorities[0] < max(priorities):
                priorities.append(priorities.pop(0))
                location -= 1 # 맨 앞 요소가 뒤로 이동하여 위치 -1 이동
            else:
                priorities.pop(0)
                location -= 1
                answer += 1
        
    return answer

# def solution(priorities, location):
#     queue =  [(i, p) for i, p in enumerate(priorities)]
#     answer = 0

#     while True:
#         cur = queue.pop(0)
#         if any(cur[1] < q[1] for q in queue):
#             queue.append(cur)
#         else:
#             answer += 1
#             if cur[0] == location:
#                 return answer

File: Programmers/Lv.2.Kotlin/프린터.kt
import java.util.*

class Solution {
    fun solution(priorities: IntArray, location: Int): Int {
        val waiting: Queue<Pair<Int, Int>> = LinkedList()
        val ready: Queue<Pair<Int, Int>> = LinkedList()
        
        priorities.forEachIndexed { index, priority ->
            waiting.add(index to priority)
        }
        
        while (waiting.isNotEmpty()) {
            if (canPrint(waiting)) {
                ready.add(waiting.poll())
            } else {
                waiting.add(waiting.poll())
            }
        }
        
        return ready.indexOf(ready.filter { it.first == location }.first()) + 1
    }
    
    fun canPrint(q: Queue<Pair<Int, Int>>): Boolean {
        var p = 0
        
        q.forEachIndexed { index, pair ->
            if (index == 0) {
                p = pair.second
            } else {
                if (p < pair.second) return false
            }
        }
        return true
    }

    // fun solution(priorities: IntArray, location: Int): Int {
    //         var printerQueue = ArrayDeque<Pair<Int,Int>>()
    //     priorities.forEachIndexed{index, i ->
    //         printerQueue.offer(Pair(index,i))
    //     }

    //     var count = 0
    //     while (!printerQueue.isEmpty()){
    //         var first = printerQueue.poll()

    //         if(printerQueue.filter { first.second < it.second }.size > 0){
    //             printerQueue.offer(first)
    //         }else{
    //             count++
    //             if(first.first == location) return count
    //         }
    //     }
    //     return 0
    // }
}